### PR TITLE
add p6spy driver support to ShardingJDBC infra

### LINKED ISSUE #31782: ShardingJDBC is not supported p6spy driver，add p6spy driver support to ShardingJDBC infra
## Bug Report

**For English only**, other languages will not accept.

Before report a bug, make sure you have:

- Searched open and closed [GitHub issues](https://github.com/apache/shardingsphere/issues).
- Read documentation: [ShardingSphere Doc](https://shardingsphere.apache.org/document/current/en/overview).

Please pay attention on issues you submitted, because we maybe need more details. 
If no response anymore and we cannot reproduce it on current information, we will **close it**.

Please answer these questions before submitting your issue. Thanks!

### Which version of ShardingSphere did you use?
5.5.0

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?
ShardingSphere-JDBC

### Expected behavior
can normal connection to the database

### Actual behavior
org.apache.shardingsphere.infra.database.core.exception.UnsupportedStorageTypeException: Unsupported storage type of URL 'jdbc:p6spy:mysql://xxx:3306?autoReconnect=true&useUnicode=true&characterEncoding=UTF-8'.

### Reason analyze (If you can)
ShadingJDBC jbdcUrl prefix not supproted jdbc:p6spy:mysql://

### Example codes for reproduce this issue (such as a github link).
```yaml
dataSources:
  db_test:
    dataSourceClassName: org.apache.tomcat.jdbc.pool.DataSource
    driverClassName: com.p6spy.engine.spy.P6SpyDriver
    url: jdbc:p6spy:mysql://xxx:3306?autoReconnect=true&useUnicode=true&characterEncoding=UTF-8
    username: u_dev
    password: xxx
```

### PR DESCRIPTION
**Pull request by lcjing**
Fixes #31782 .

Changes proposed in this pull request:
  -

---

Before committing this PR, I'm sure that I have checked the following options:
- [ ] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [ ] I have self-reviewed the commit code.
- [ ] I have (or in comment I request) added corresponding labels for the pull request.
- [ ] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [ ] I have made corresponding changes to the documentation.
- [ ] I have added corresponding unit tests for my changes.

### CODE REVIEW

File: infra/database/type/p6spy/src/main/java/org/apache/shardingsphere/infra/database/p6spy/type/P6spyMySQLDatabaseType.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import java.util.Optional;
+
+/**
+ * Database type of MySQL under P6Spy.
+ */
+public class P6spyMySQLDatabaseType implements DatabaseType {
+    @Override
+    public Collection<String> getJdbcUrlPrefixes() {
+        return Collections.singletonList("jdbc:p6spy:mysql:");
+    }
```
Comment (linghengqian) [inline infra/database/type/p6spy/src/main/java/org/apache/shardingsphere/infra/database/p6spy/type/P6spyMySQLDatabaseType.java:34]: - I don't see any testcontainers module for p6spy at https://github.com/testcontainers/testcontainers-java . Is the `org.testcontainers.containers.JdbcDatabaseContainerProvider` SPI implementation corresponding to the jdbcUrl prefix open-source?

**Review by linghengqian — COMMENTED** [review]
- The PR did not pass the code formatting validation. You can execute first `./mvnw spotless:apply -Pcheck -T1C`. Then execute `./mvnw checkstyle:check -Pcheck -T1C` to manually adjust to repair the CI.

File: infra/database/type/testcontainers/src/main/java/org/apache/shardingsphere/infra/database/testcontainers/type/TcP6spyMySQLDatabaseType.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * Database type of MariaDB in testcontainers-java.
+ */
+public final class TcP6spyMySQLDatabaseType implements TestcontainersDatabaseType {
+    
+    @Override
+    public Collection<String> getJdbcUrlPrefixes() {
+        return Collections.singleton("jdbc:tc:p6spy:mysql:");
+    }
```
Comment (linghengqian) [inline infra/database/type/testcontainers/src/main/java/org/apache/shardingsphere/infra/database/testcontainers/type/TcP6spyMySQLDatabaseType.java:35]: - I don't see any testcontainers module for p6spy at https://github.com/testcontainers/testcontainers-java . Is the `org.testcontainers.containers.JdbcDatabaseContainerProvider` SPI implementation corresponding to the jdbcUrl prefix open-source?

File: infra/database/type/p6spy/pom.xml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+            <groupId>org.apache.shardingsphere</groupId>
+            <artifactId>shardingsphere-infra-database-core</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+
+        <dependency>
+            <groupId>org.apache.shardingsphere</groupId>
+            <artifactId>shardingsphere-infra-database-mysql</artifactId>
+            <version>${project.version}</version>
+        </dependency>
```
[the remark this answers is not in the document]
Comment (lcjing) [reply infra/database/type/p6spy/pom.xml:40]: Oh, It seems so, I removed this dependency

File: infra/database/type/testcontainers/src/main/java/org/apache/shardingsphere/infra/database/testcontainers/type/TcP6spyMySQLDatabaseType.java (hunk 2 above)
Comment (lcjing) [reply infra/database/type/testcontainers/src/main/java/org/apache/shardingsphere/infra/database/testcontainers/type/TcP6spyMySQLDatabaseType.java:35]: Thanks for pointing this out. I didn't understand the definition of the testcontainers module and thought it was a test case.  Maybe I shouldn't modify this module?

Comment (linghengqian) [reply infra/database/type/testcontainers/src/main/java/org/apache/shardingsphere/infra/database/testcontainers/type/TcP6spyMySQLDatabaseType.java:35]: - The testcontainers module of shardingsphere infra corresponds to several SPI implementations of `org.testcontainers.containers.JdbcDatabaseContainerProvider`. For example, https://github.com/testcontainers/testcontainers-java/blob/main/modules%2Fmysql%2Fsrc%2Fmain%2Fjava%2Forg%2Ftestcontainers%2Fcontainers%2FMySQLContainerProvider.java . 

- If there is no corresponding SPI implementation class, then there is actually no need to modify the infra testcontainers module.

Comment (lcjing) [reply infra/database/type/testcontainers/src/main/java/org/apache/shardingsphere/infra/database/testcontainers/type/TcP6spyMySQLDatabaseType.java:35]: OK, thanks, I understand. I will resubmit the changes.

**Review by linghengqian — APPROVED** [review]
LGTM!

File: infra/database/type/testcontainers/src/main/java/org/apache/shardingsphere/infra/database/testcontainers/type/TcP6spyMySQLDatabaseType.java (hunk 2 above)
Comment (linghengqian) [reply infra/database/type/testcontainers/src/main/java/org/apache/shardingsphere/infra/database/testcontainers/type/TcP6spyMySQLDatabaseType.java]: > OK, thanks, I understand. I will resubmit the changes.

- Due to the ongoing controversy on the testcontainers side of the PR, I added documentation for this optional module at #33707 and https://shardingsphere.apache.org/document/current/cn/user-manual/shardingsphere-jdbc/optional-plugins/testcontainers/ .

**Pull request merged**